src/jit/gentree.cpp

```diff
@@ -8817,6 +8821,20 @@ unsigned GenTree::NumChildren()
             }
             return childCount;
         }
+#ifdef FEATURE_HW_INTRINSICS
+        // GT_HWIntrinsic require special handling in case of SetZeroVectorNNN functions which do not take arg
+        if (OperGet() == GT_HWIntrinsic)
```

Comment:
We could probably just return the number of children from the HWIntrinsic table....

Reply:
That wont work, since fixed immediate have 2 children in table but 3 in IR

Reply:
`max(numArgs, 2)`

Reply:
There are 1 arg HWIntrinsics which should be handled as well ... the code after seems to do it correctly except for HWIntrinsics which have 1st argument in the form of `GenTreeArgList` where they will be handled as 1 arg instead of expanding args ... perhaps we should include them as special case as well?

Reply:
Yes. Which is why just doing `max(numArgs, 2)` should work

Reply:
My understanding is that we can handle HWIntrinsics in switch table in the following way:
1. All zero argument intrinsics return 0.
2. All legitimate 1 argument intrinsics return 1.
3. All other intrinsics return `max(numArgs, 2)`.
We take number of arguments from table. Is that correct?

Reply:
Sorry, it should be `min(numArgs, 2)` (not `max`) in all cases (with, possibly an assert that `numArgs` isn't negative). That will return 0 for zero args, 1 for 1 arg, and 2 for 2+ args.

Reply:
How about legitimate 3 args intrinsics? Ex. `SSE41_BlendVariable`, perhaps we should return 3 ...

Reply:
Okay, I looked at the consumers of `NumChildren` and this more in depth.

We want to return 0 for any HWIntrinsic operations for which `gtOp1` and `gtOp2` are `nullptr`
We want to return 1 for any HWIntrinsic operations for which just `gtOp2` is `nullptr`
We want to return 2 for any HWIntrinsic operations for which neither are `nullptr`

There are several consumers of `NumChildren` which may (or may not) need to be updated to handle the case where `NumChildren()` is 1, and where the child is an arg list.

Reply:
Essentially it means that want to special case only HWIntrinsic operations for which gtOp1 and gtOp2 are nullptr, since existing code will handle properly all other cases and special casing in consumers will be based on HWIntrinsic operation check. Is that what should be implemented here?

Reply:
Yes

Reply:
Yes